### PR TITLE
<WIP> feat: support writing domain metadata (2/2)

### PR DESCRIPTION
**Pull request by MannDP**

### CODE REVIEW

File: kernel/src/actions/mod.rs [hunk 1]
```diff
@@ -885,9 +885,29 @@ impl DomainMetadata {
     // returns true if the domain metadata is an system-controlled domain (all domains that start
     // with "delta.")
     #[allow(unused)]
-    fn is_internal(&self) -> bool {
+    pub(crate) fn is_internal(&self) -> bool {
```
Comment (MannDP) [inline kernel/src/actions/mod.rs:984]: Note to reviewers: please only view the last commit in this PR. It is logically stacked on top of https://github.com/delta-io/delta-kernel-rs/pull/1274/commits.

File: kernel/src/actions/mod.rs [hunk 2]
```diff
@@ -939,6 +939,15 @@ impl DomainMetadata {
         }
     }
 
+    // Create a new DomainMetadata action to remove a domain.
+    pub(crate) fn remove(domain: String) -> Self {
+        Self {
+            domain,
+            configuration: String::new(),
```
Comment (zachschuermann) [inline kernel/src/actions/mod.rs]: is this how existing implementations do removes? I kinda figured it would retain the old configuration but haven't dug into this yet

File: kernel/src/transaction/mod.rs [hunk 3]
```diff
@@ -277,6 +277,20 @@ impl Transaction {
...
+    /// This creates a tombstone to logically delete the specified domain. We don't check
+    /// if the domain metadata exists in the log, the remove action is simply a tombstone record.
+    /// Note that each domain can only appear once per transaction. That is, multiple operations
+    /// on the same domain are disallowed in a single transaction, as well as setting and removing
+    /// the same domain in a single transaction. If a duplicate domain is included, the `commit` will
+    /// fail (that is, we don't eagerly check domain validity here).
+    /// Removing metadata for multiple distinct domains is allowed.
+    pub fn with_domain_metadata_removed(mut self, domain: String) -> Self {
+        self.domain_metadatas
+            .push(DomainMetadata::remove(domain.clone()));
```
Comment (zachschuermann) [inline kernel/src/transaction/mod.rs]:
```suggestion
            .push(DomainMetadata::remove(domain));
```

File: kernel/src/actions/mod.rs (hunk 2 above)
Comment (MannDP) [reply kernel/src/actions/mod.rs]: great call out! i went back and looked at the protocol again. "Writers should preserve an accurate pre-image of the configuration" -> so we should keep it around actually.

Comment (lbhm) [reply kernel/src/actions/mod.rs]: Retaining the old configuration would require log replay though, right? We could accept `configuration: Option<String>` as an additional parameter to let the caller provide the old value if already known, but TBH I'm not sure if that improves the current solution.

File: kernel/src/transaction/mod.rs [hunk 4]
```diff
@@ -277,6 +277,20 @@ impl Transaction {
         self
     }
 
+    /// Remove domain metadata from the Delta log.
+    /// This creates a tombstone to logically delete the specified domain. We don't check
```
Comment (lbhm) [inline kernel/src/transaction/mod.rs]: nit: IMO it is good style to have a paragraph break (i.e., empty line) after the one-line summary.

File: kernel/src/actions/mod.rs [hunk 5]
```diff
@@ -939,6 +939,15 @@ impl DomainMetadata {
         }
     }
 
+    // Create a new DomainMetadata action to remove a domain.
+    pub(crate) fn remove(domain: String) -> Self {
```
Comment (lbhm) [inline kernel/src/actions/mod.rs]: I feel like `remove` is a somewhat misleading name for a method that is actually a constructor. I don't have a strong opinion here, but maybe something like `new_tombstone` communicates the function's intent more clearly?

File: kernel/tests/write.rs [hunk 6]
```diff
@@ -1382,3 +1382,198 @@ async fn test_set_domain_metadata_unsupported_writer_feature(
...
+        .into_iter()
+        .try_collect()?;
+
+    let domain_action = actions
+        .iter()
+        .find(|v| v.get("domainMetadata").is_some())
+        .unwrap();
+    assert_eq!(domain_action["domainMetadata"]["domain"], "app.deprecated");
+    assert_eq!(domain_action["domainMetadata"]["configuration"], "");
+    assert_eq!(domain_action["domainMetadata"]["removed"], true);
```
Comment (lbhm) [inline kernel/tests/write.rs]: Doesn't your clippy complain that this should be 
```rust
assert!(domain_action["domainMetadata"]["removed"]);
```
?

File: kernel/tests/write.rs
Comment (lbhm) [inline kernel/tests/write.rs:1]: Should we maybe also test that `domain_metadata::domain_metadata_configuration` correctly returns `None` when called with a domain that has a tombstone?

Comment (MannDP) [reply kernel/tests/write.rs:1]: Good call out. The last test verifies this case. It sets the domain metadata in txn1 then removes it in txn2. Then we assert that:
1. The log contains the tombstone record
2. Calling get domain metadata for that domain returns `None`

File: kernel/src/transaction/mod.rs (hunk 4 above)
Comment (MannDP) [reply kernel/src/transaction/mod.rs]: hmm, I'm not following what you mean. Do you mean the "We don't check" should start on the next line perhaps?

File: kernel/tests/write.rs (hunk 6 above)
Comment (MannDP) [reply kernel/tests/write.rs]: I think we cannot do it that way because `actions: Vec<serde_json::Value>`.
```
cannot apply unary operator `!` to type `Value`
cannot apply unary operator `!`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20[0]?0#file:///home/mann.patel/delta-kernel-rs/kernel/tests/write.rs)
mod.rs(116, 1): the foreign item type `Value` doesn't implement `std::ops::Not`
```

File: kernel/src/actions/mod.rs (hunk 2 above)
Comment (MannDP) [reply kernel/src/actions/mod.rs]: You're right that it requires log replay. But as long as we do that once it should be okay. I checked the Java version of the delta kernel for inspiration and they follow this pattern where remove takes only the `domain` and they internally determine the pre-image configuration value to preserve. Code ptr: https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java#L230C15-L230C35.

The other thing is if we let users pass it in, we likely _should_ validate that the pre-image they provide is correct. Which would require log replay anyways.

File: kernel/tests/write.rs (hunk 6 above)
Comment (lbhm) [reply kernel/tests/write.rs]: Ah I see

File: kernel/src/transaction/mod.rs (hunk 4 above)
Comment (lbhm) [reply kernel/src/transaction/mod.rs]: I meant
```rust
/// Remove domain metadata from the Delta log.
///
/// This creates a tombstone to logically delete the specified domain. We don't check
```

See [this](https://rust-lang.github.io/rust-clippy/master/index.html#too_long_first_doc_paragraph) clippy lint for reference.

File: kernel/src/transaction/mod.rs [hunk 7]
```diff
@@ -295,32 +315,58 @@ impl Transaction {
...
+                    existing_domains.get(dm.domain()).map(|existing| {
+                        DomainMetadata::remove(
+                            dm.domain().to_string(),
+                            existing.configuration().to_string(),
+                        )
+                    })
+                } else {
+                    Some(dm.clone())
+                }
+            });
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/mod.rs:437]: We can consume the domain_metadatas and avoid the clone since it seems like we don't use it anymore
```suggestion
        let user_domains = self
            .domain_metadatas
            .into_iter()
            .filter_map(move |dm: DomainMetadata| {
                if !dm.is_removed() {
                    return Some(dm);
                }
                existing_domains.get(dm.domain()).map(|existing| {
                    DomainMetadata::remove(
                        dm.domain().to_string(),
                        existing.configuration().to_string(),
                    )
                })
            });
```

File: kernel/src/transaction/mod.rs [hunk 8]
```diff
@@ -295,32 +315,58 @@ impl Transaction {
...
+                has_removals = true;
+            }
         }
 
+        // fetch previous configuration values (requires log replay)
+        let existing_domains = if has_removals {
+            scan_domain_metadatas(self.read_snapshot.log_segment(), None, engine)?
+        } else {
+            HashMap::new()
+        };
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/mod.rs:421]: I'm not convinced we need to verify that the domain metadata exists. After all, Delta never checks that a removed file is actually present. Is this something that we need to do? If so, can you put a comment explaining why we need to read the entire log and set the removed configuration?

File: kernel/src/transaction/mod.rs [hunk 9]
```diff
@@ -277,7 +277,27 @@ impl Transaction {
...
+    /// If the domain does not exist in the Delta log, this is a no-op.
+    /// Note that each domain can only appear once per transaction. That is, multiple operations
+    /// on the same domain are disallowed in a single transaction, as well as setting and removing
+    /// the same domain in a single transaction. If a duplicate domain is included, the `commit` will
+    /// fail (that is, we don't eagerly check domain validity here).
+    /// Removing metadata for multiple distinct domains is allowed.
+    pub fn with_domain_metadata_removed(mut self, domain: String) -> Self {
+        // actual configuration value determined during commit
+        self.domain_metadatas
+            .push(DomainMetadata::remove(domain, String::new()));
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/mod.rs:367]: Instead of having the dummy data and checking below whether we have a removal, why not maintain a separate `removed_domains: HashSet<String>` that we fill in after the log replay?

Comment (OussamaSaoudi) [reply kernel/src/transaction/mod.rs:367]: Note: only do this if we actually do need to preserve the removed domain metadata's configuration. See my comment here: https://github.com/delta-io/delta-kernel-rs/pull/1275/files#r2400384116

File: kernel/src/transaction/mod.rs [hunk 10]
```diff
@@ -295,32 +315,58 @@ impl Transaction {
...
             }
-            if !domains.insert(domain_metadata.domain()) {
+
+            if !seen_domains.insert(dm.domain()) {
                 return Err(Error::Generic(format!(
                     "Metadata for domain {} already specified in this transaction",
-                    domain_metadata.domain()
+                    dm.domain()
                 )));
             }
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/mod.rs:409]: These feel like checks that should be done in `with_domain_metadata`:
* `if dm.is_internal()`
*  `if !seen_domains.insert(dm.domain())`

Generally I prefer to fail early. Feel free to make this a followup issue.

File: kernel/src/transaction/mod.rs [hunk 11]
```diff
@@ -295,32 +315,58 @@ impl Transaction {
...
+
+            if dm.is_removed() {
+                has_removals = true;
+            }
```
Comment (OussamaSaoudi) [inline kernel/src/transaction/mod.rs:413]: Linearly checking domain metadatas can be avoided by maintaining a separate hashmap. See [here](https://github.com/delta-io/delta-kernel-rs/pull/1275/files#r2400385771)

File: kernel/src/transaction/mod.rs (hunk 7 above)
Comment (OussamaSaoudi) [reply kernel/src/transaction/mod.rs:437]: If we decide to use `removed_domains` this becomes:
We can consume the domain_metadatas and avoid the clone since it seems like we don't use it anymore
```suggestion
        let user_domains = self
            .removed_domains
            .iter()
            .map(move |domain_name: &str| {
                existing_domains.get(domain_name).map(|existing| {
                    DomainMetadata::remove(
                        dm.domain().to_string(),
                        existing.configuration().to_string(),
                    )
                })
            })
            .chain(domain_metadatas);
```

**Review by OussamaSaoudi — CHANGES_REQUESTED** [review]
Nice tests! Just some code reshuffling then we're good to go 👍

File: kernel/src/transaction/mod.rs (hunk 9 above)
Comment (MannDP) [reply kernel/src/transaction/mod.rs:367]: The Delta spec requires it for configuration values. See here: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#:~:text=When%20true%2C%20the%20action%20serves%20as%20a%20tombstone%20to%20logically%20delete%20a%20metadata%20domain.%20Writers%20should%20preserve%20an%20accurate%20pre%2Dimage%20of%20the%20configuration..

See this thread for the prior discussion on this: https://github.com/delta-io/delta-kernel-rs/pull/1275#discussion_r2379752520.

File: kernel/src/transaction/mod.rs (hunk 10 above)
Comment (MannDP) [reply kernel/src/transaction/mod.rs:409]: I generally agree with the fail fast principle, but this case warrants an exception imo.

We're treating these methods as builder methods (return `Self`, even though there is no explicit `build` call) so we should present a fluent builder API. If we do the checking of the error conditions at this point, then we need to return `DeltaResult<Self>`, which is an awkward interface for both Rust (chaining requires try operator), but particularly for engines across the FFI boundary.

wdyt?

File: kernel/src/transaction/mod.rs (hunk 11 above)
Comment (MannDP) [reply kernel/src/transaction/mod.rs:413]: Let's bottom out on deferred error checking or not above. Because if we choose to defer error checking, then we cannot make it a `HashMap`. If we decide to not defer, then will refactor accordingly.

File: kernel/src/transaction/mod.rs (hunk 8 above)
Comment (MannDP) [reply kernel/src/transaction/mod.rs:421]: added a comment (that the Delta spec requires it). let's bottom out on this in one of the above threads.

**Review by MannDP — COMMENTED** [review]
tysm for the review! wanted to align on 2 high-level things so added thoughts on.
1. preserve pre-image of removed configuration
2. error checking, fail-fast or in `commit`

File: kernel/src/transaction/mod.rs (hunk 7 above)
Comment (MannDP) [reply kernel/src/transaction/mod.rs:437]: will probably apply suggestion 1 or 2 depending on our takeaways from the other threads :)

File: kernel/src/transaction/mod.rs (hunk 8 above)
Comment (OussamaSaoudi) [reply kernel/src/transaction/mod.rs:421]: Just saw the other thread. Thanks!

File: kernel/src/transaction/mod.rs (hunk 10 above)
Comment (OussamaSaoudi) [reply kernel/src/transaction/mod.rs:409]: > which is an awkward interface for both Rust (chaining requires try operator)

I was willing to deal with an awkward interface

> but particularly for engines across the FFI boundary. 

But this is a really good point! GJ

File: kernel/src/actions/mod.rs (hunk 2 above)
Comment (OussamaSaoudi) [reply kernel/src/actions/mod.rs]: Yeah I think we should compute pre-image ourselves since as you point out, we'd have to check anyway.

File: kernel/src/transaction/mod.rs (hunk 7 above)
Comment (OussamaSaoudi) [reply kernel/src/transaction/mod.rs:437]: oof I'm also realizing that because "generate_domain_metadata" can be called multiple times, the `into_iter` doesn't work. Also given the deferred error handling, I think removes can remain as they are for now.

**Comment by OussamaSaoudi** [comment]
I think I'm okay with deferred error handling for now. I think the flow/error handling could be improved a bit as we move to splitting the Transaction type into different states.

Regarding the pre-image of removed configuration, I think we should compute it ourselves since we're gonna have to do it anyway.

File: kernel/src/transaction/mod.rs (hunk 9 above)
Comment (zachschuermann) [inline kernel/src/transaction/mod.rs:367]: I think we seem to be conflating the API communicating domain metadata additions/removals with the underlying DomainMetadata itself.

Can we simplify and make our lives easier below if we just take a set of domains to remove? Then we (1) don't have the 'fill-in-later' behavior here, and (2) immediately know if we have removals

I suppose we might need to still have something that knows our global set of domains since it's a logic error to add/remove in the same transaciton..

Comment (MannDP) [reply kernel/src/transaction/mod.rs:367]: hmm this is interesting. if we have a separate state variable for removals, we can make it something like `HashSet<String>` instead. then we immediately know whether there are removals or not (point 2).

regarding point 1, the "fill-in" behavior, we _can_ do log replay and store that value in a `OnceCell` or similar on the first removal. Subsequent removals can then immediately look up the configuration value to preserve. But, I think now we are adding class level state for something that could have been totally local in `commit`.

wdyt about adding a separate state variable as `HashSet<String>` for removals @zachschuermann?

Comment (MannDP) [reply kernel/src/transaction/mod.rs:367]: resolved offline; will merge pr as-is. (can cut a follow-ups as this is internal state only - no external API)
